## Integrantes

- **Estudiante 1:** Juan David Duarte Yara - 202215070
- **Estudiante 2:** Luisa Gabriela Hernández Pérez - 202114093
- **Estudiante 3:** Juan Sebastian Sanchez - 202121498
- Todos los estudiantes aportaron en las partes generales del notebook y en las diapositivas de coclusiones.



## 0. Carga de los datos. 

### 0.1. Importación de librerías 

En las siguientes líneas de código se importan las librerías y herramientas necesarias para desarrollar el caso de uso.

In [159]:
# Importaciones de bibliotecas
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import sys
import re
import string
import unicodedata

# Importaciones de joblib
from joblib import dump, load

# Importaciones de sklearn
from sklearn.base import BaseEstimator, RegressorMixin, ClassifierMixin
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, FunctionTransformer
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, classification_report, confusion_matrix
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer, HashingVectorizer
from sklearn.svm import SVC
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import BernoulliNB
from sklearn.impute import SimpleImputer
from sklearn import tree
from sklearn.metrics import confusion_matrix, classification_report, precision_score, recall_score, f1_score, accuracy_score
from sklearn.metrics import ConfusionMatrixDisplay

# Clasificacion
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from imblearn.over_sampling import SMOTE
from sklearn.datasets import make_classification

# Importaciones de statsmodels
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.stats.diagnostic import het_breuschpagan, het_white

# Importaciones de nltk
import nltk
from nltk import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import LancasterStemmer, WordNetLemmatizer, PorterStemmer, SnowballStemmer

# Instalación de paquetes adicionales
import contractions

# Para búsqueda de hiperparámetros
from sklearn.model_selection import GridSearchCV

# Para la validación cruzada
from sklearn.model_selection import KFold

### 0.2. Lectura de archivos
A través de la librería **pandas** podemos realizar la carga de datos desde diferentes fuentes de información, en este caso se realizará la carga de un archivo plano csv (archivo separado por comas).

In [160]:
# Se cargan los datos. 
datatrain=pd.read_csv('fake_news_spanish.csv', sep=';', encoding = 'utf-8')
datatest = pd.read_csv('fake_news_test.csv', sep=';', encoding = 'utf-8')

Podemos ver que hay 57,063 filas y 5 columnas de atributos en el conjunto de entrenamiento

In [161]:
datatrain.shape

(57063, 5)

Podemos ver que hay 1000 filas y 4 columnas de atributos en el conjunto de prueba

In [162]:
datatest.shape

(1000, 4)

Veamos un pequeño sample de los datos

In [163]:
datatrain.sample(10)

ID  Label                                             Titulo  \
37379  ID      1  Couso: 'El neofascismo sólo está logrando enfr...   
39889  ID      1  Anticapitalistas rechazan un gobierno de coali...   
52199  ID      0  Feijóo presume de la estabilidad en su Gobiern...   
45003  ID      1  Miguel Urbán: 'El mayor riesgo de atentado en ...   
11431  ID      0  ¿Es delito hacer chistes o bromas sobre Alfons...   
39274  ID      1  Aguado admite que los menús de Ayuso no son sa...   
8336   ID      1  Programas sin números ni cuentas: los 'brindis...   
32240  ID      1  Más de 3.500 jueces y unos 1.300 fiscales secu...   
2600   ID      1  Ione Belarra toma las riendas de Podemos mient...   
45302  ID      0  El Iniciativa vers per Catalunya andaluz dice ...   

                                             Descripcion       Fecha  
37379  El exeurodiputado de IU quiere volver a Brusel...  18/05/2019  
39889  Los modelos verticales han contribuido al vaci...  31/05/2019  
52199  En plena tormenta política por las mociones de...  11/03/2021  
45003  Podríamos decir que desde el atentado en Norue...  11/12/2022  
11431  El Tribunal Supremo confirmó el pasado viernes...  06/02/2018  
39274  A nadie se le escapa que de continuar más días...  06/05/2020  
8336   Los partidos vuelven a hacer una precampaña en...  29/10/2019  
32240  El 65% de los jueces y el 52% de los fiscales ...  21/05/2018  
2600   La renovación en el liderazgo de la formación ...  13/06/2021  
45302  El vicesecretario del Iniciativa vers per Cata...  22/09/2020

Revisemos que el tipo de dato sea el esperado

In [164]:
datatrain.dtypes

ID             object
Label           int64
Titulo         object
Descripcion    object
Fecha          object
dtype: object

# 2. Entendimiento de los datos

## 2.1 Analisis descriptivo

Podemos ver que hay textos repetidos

In [165]:
datatrain["Descripcion"].describe()


count                                                 57063
unique                                                49638
top       La Fundación En Acción ha participado con Públ...
freq                                                      7
Name: Descripcion, dtype: object

Es útil saber que tan largos son los textos

In [166]:
datatrain["Descripcion"].apply(len).describe()

count    57063.000000
mean       243.519724
std        274.479568
min         33.000000
25%        171.000000
50%        221.000000
75%        267.000000
max      14153.000000
Name: Descripcion, dtype: float64

Los textos tienen una longitud entre 33 y 14,153 caracteres, con una media de 243 caracteres. 

Queremos ver ahora la columna "label"

In [167]:
datatrain["Label"].describe()

count    57063.000000
mean         0.581077
std          0.493387
min          0.000000
25%          0.000000
50%          1.000000
75%          1.000000
max          1.000000
Name: Label, dtype: float64

Podemos ver que efectivamente la columna "Label" tiene valores entre 0 y 1.

Nos gustaría saber la proporción que hay entre noticias falsas y las que no

In [168]:
datatrain["Label"].value_counts()

Label
1    33158
0    23905
Name: count, dtype: int64

Podemos ver que los datos están ligeramente desbalanceados

## 2.2 Calidad de datos

In [169]:
# Es recomendable que todos los pasos de limpieza y preparación se realicen sobre otro archivo.
data_t = datatrain.copy()

### Unicidad

Revisemos los duplicados por título 

In [170]:
data_t.loc[datatrain.duplicated(subset= "Titulo", keep = False)].sort_values("Titulo").head(10)

ID  Label                                             Titulo  \
53500  ID      1  "La quimioterapia mata a la gente y no al cánc...   
27107  ID      1  "La quimioterapia mata a la gente y no al cánc...   
44905  ID      1  #DespuésDeAplaudir queda mucho por hacer y rei...   
41264  ID      0  #DespuésDeAplaudir queda mucho por hacer y rei...   
9017   ID      1  '11 millones de euros por teléfono y por Whats...   
25449  ID      0  '11 millones de euros por teléfono y por Whats...   
42708  ID      1  'A este hay que matarlo'... Las perlas de las ...   
41717  ID      0  'A este hay que matarlo'... Las perlas de las ...   
26334  ID      1  'A mi madre la llevaron al sótano de la reside...   
7823   ID      0  'A mi madre la llevaron al sótano de la reside...   

                                             Descripcion       Fecha  
53500  Señor Presidente recibiremos sus tropas con lo...  01/11/2018  
27107   Estamos en la semana decisiva. Los expertos a...  10/08/2021  
44905  Acto a favor de la Sanidad pública y universal...  28/05/2020  
41264  Acto a favor de la Sanidad pública y universal...  28/05/2020  
9017   Los comisionistas del escándalo de las masacri...  26/04/2022  
25449  Los comisionistas del escándalo de las masacri...  26/04/2022  
42708  El comisario relata el robo de información al ...  30/09/2020  
41717  El comisario relata el robo de información al ...  30/09/2020  
26334  Comienzan a declarar en el juzgado los familia...  22/02/2021  
7823   Comienzan a declarar en el juzgado los familia...  22/02/2021

Revisemos ahora los duplicados por descripción 

In [171]:
data_t.loc[datatrain.duplicated(subset= "Descripcion", keep = False)].sort_values("Descripcion").head(10)

ID  Label                                             Titulo  \
10367  ID      0  Inma Nieto: 'José Manuel Franco me decía hace ...   
51799  ID      1  Edmundo Bal: 'Adrián Vázquez me decía hace poc...   
38903  ID      1  Un manifiesto llama a que Xavier Domènech lide...   
31583  ID      0  Un manifiesto llama a que Adriana Lastra lider...   
27018  ID      1  ¿Rey franquista o padre de la democracia?: luc...   
14590  ID      0  ¿Rey franquista o padre de la democracia?: luc...   
20779  ID      1  Los médicos de Atención Primaria en Madrid se ...   
49171  ID      0  Los médicos de Atención Primaria en Madrid se ...   
27698  ID      1  Procesados Sito Miñanco y Gonzalo Boye por bla...   
40787  ID      0  Procesados Sito Miñanco y Feijoo por blanqueo ...   

                                             Descripcion       Fecha  
10367  (Risas) Sí...lo que pasa es que esto viene ya ...  03/01/2023  
51799  (Risas) Sí...lo que pasa es que esto viene ya ...  03/01/2023  
38903  134 miembros destacados de la formación catala...  02/03/2018  
31583  134 miembros destacados de la formación catala...  02/03/2018  
27018  43 años y medio después de jurar los principio...  01/06/2019  
14590  43 años y medio después de jurar los principio...  01/06/2019  
20779  5.000 sanitarios de los Centros de Salud de la...  10/03/2021  
49171  5.000 sanitarios de los Centros de Salud de la...  10/03/2021  
27698  A Boye, el abogado de Puigdemont, se le relaci...  18/12/2020  
40787  A Boye, el abogado de Puigdemont, se le relaci...  18/12/2020

In [172]:
data_t.duplicated(keep=False).sum()

875

Revisando el archivo pudimos notar que realmente los anteriores datos no son duplicados, sino que tienen ligeras diferencias entre su contenido, reafirmando que uno es una fake news y la otra no.

Ahora revisemos cuando la columna "Titulo", "Descripcción","Fecha" y "Label" son exactamente iguales.

In [173]:
data_t.loc[datatrain.duplicated(subset= ["Descripcion","Titulo"], keep = False)].sort_values("Descripcion").head(10)

ID  Label                                             Titulo  \
54191  ID      1  Madrid envía una carta a los directores de los...   
54892  ID      1  Madrid envía una carta a los directores de los...   
1551   ID      1  El legado del PP en Madrid: desigualdad, anemi...   
8627   ID      1  El legado del PP en Madrid: desigualdad, anemi...   
16683  ID      1  Ayuso recurre a las descalificaciones para def...   
18999  ID      1  Ayuso recurre a las descalificaciones para def...   
55074  ID      1  ERC y JxCat se reúnen en la cárcel de Lledoner...   
27758  ID      1  ERC y JxCat se reúnen en la cárcel de Lledoner...   
36567  ID      1  Otra amenaza de muerte sin una contundente con...   
29871  ID      1  Otra amenaza de muerte sin una contundente con...   

                                             Descripcion       Fecha  
54191  A cinco días del 4M, el director general de Ed...  30/04/2021  
54892  A cinco días del 4M, el director general de Ed...  30/04/2021  
1551   A lo largo de la campaña se hablará más en cla...  19/04/2021  
8627   A lo largo de la campaña se hablará más en cla...  19/04/2021  
16683  A lo largo del debate la dirigente conservador...  21/04/2021  
18999  A lo largo del debate la dirigente conservador...  21/04/2021  
55074  A menos de un mes para agotar el plazo para in...  27/04/2021  
27758  A menos de un mes para agotar el plazo para in...  27/04/2021  
36567  Abascal dice que la carta con amenazas a Igles...  26/04/2021  
29871  Abascal dice que la carta con amenazas a Igles...  26/04/2021

Todas estas instancias son exactamente iguales, por lo cual, son duplicados. Revisemos cuantos duplicados realmente hay.

In [174]:
data_t.duplicated(subset=['Titulo', 'Descripcion','Fecha','Label'], keep=False).sum()

875

Borremos los duplicados

In [175]:
data_t = data_t.drop_duplicates(subset=['Titulo', 'Descripcion','Fecha','Label'], keep='first')

Revisemos si el tamaño de los datos cambió. El cambio debe ser consistente con la cantidad de duplicados eliminados

In [176]:
datatrain.shape

(57063, 5)

In [177]:
data_t.shape

(56618, 5)

Revisemos si hay duplicados

In [178]:
data_t.duplicated(subset=['Titulo', 'Descripcion','Fecha','Label'], keep=False).sum()

0

### Completitud

In [179]:
# Se observa que hay ausencias en la columna titulo
data_t.isnull().sum()

ID              0
Label           0
Titulo         16
Descripcion     0
Fecha           0
dtype: int64

In [180]:
# Como solo son 16 instancias de más de 57,000 podemos borrar estos datos
data_t = data_t.dropna(subset=["Titulo"])

In [181]:
# Se observa que ya no hay ausencias en la columna titulo
data_t.isnull().sum()

ID             0
Label          0
Titulo         0
Descripcion    0
Fecha          0
dtype: int64

### Validez

Podemos confirmar que solo existen los labels de fake news y no fake news.

In [182]:
pd.DataFrame(data_t["Label"].unique())

0
0  1
1  0

### Consistencia

Revisemos el formato de las fechas

In [183]:
data_t["Fecha"].isna().sum()


0

In [184]:
data_t['Nueva_Fecha'] = pd.to_datetime(data_t['Fecha'], format="%d/%m/%Y", errors ='coerce')
data_t['Nueva_Fecha'].sample(10)

31332   2020-02-21
30972   2023-01-26
4530    2020-01-03
14927   2020-06-16
41641   2021-02-08
15398   2018-06-15
48385   2023-03-06
26423   2022-02-07
16827   2020-12-04
38115   2018-06-04
Name: Nueva_Fecha, dtype: datetime64[ns]

In [185]:
# Miramos los registros que no pudieron convertirse a DateTime. Vemos que todos se convirtieron satisfactoriamente a datetime y no hay ningun NaT
data_t.loc[data_t.Nueva_Fecha.isna(), ['Fecha','Nueva_Fecha']] 

Empty DataFrame
Columns: [Fecha, Nueva_Fecha]
Index: []

In [187]:
#Revisemos los tipos de datos
data_t.dtypes

ID                     object
Label                   int64
Titulo                 object
Descripcion            object
Fecha                  object
Nueva_Fecha    datetime64[ns]
dtype: object